<a href="https://colab.research.google.com/github/abilashkannanv/AIML/blob/main/TrustCart_Phase1_Transaction_Risk_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd

## Task 1.1 Dataset Familiarization

In [2]:
# Load transactions.csv
transactions_df = pd.read_csv('/content/sample_data/transaction.csv')

# Load identity.csv
identity_df = pd.read_csv('/content/sample_data/identity.csv')

print('Datasets loaded successfully.')

Datasets loaded successfully.


### Analysis of `transactions.csv`

In [3]:
# Identify number of records and features for transactions_df
num_records_transactions = transactions_df.shape[0]
num_features_transactions = transactions_df.shape[1]

print(f"Number of records in transactions.csv: {num_records_transactions}")
print(f"Number of features in transactions.csv: {num_features_transactions}")

# Check for target variable (risk_flag) in transactions_df
if 'risk_flag' in transactions_df.columns:
    print("Target variable 'risk_flag' found in transactions.csv.")
else:
    print("Target variable 'risk_flag' not found in transactions.csv.")

print("\nFirst 5 rows of transactions.csv:")
display(transactions_df.head())

print("\nInformation about transactions.csv:")
transactions_df.info()

Number of records in transactions.csv: 296234
Number of features in transactions.csv: 394
Target variable 'risk_flag' not found in transactions.csv.

First 5 rows of transactions.csv:


,TransactionID,isFraud,TransactionDT,TransactionAmt,ProductCD,card1,card2,card3,card4,card5,...,V330,V331,V332,V333,V334,V335,V336,V337,V338,V339
0,2987000,0,86400,68.5,W,13926,NaN,150.0,discover,142.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2987001,0,86401,29.0,W,2755,404.0,150.0,mastercard,102.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2987002,0,86469,59.0,W,4663,490.0,150.0,visa,166.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2987003,0,86499,50.0,W,18132,567.0,150.0,mastercard,117.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2987004,0,86506,50.0,H,4497,514.0,150.0,mastercard,102.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0



Information about transactions.csv:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 296234 entries, 0 to 296233
Columns: 394 entries, TransactionID to V339
dtypes: float64(376), int64(4), object(14)
memory usage: 890.5+ MB


### Analysis of `identity.csv`

In [4]:
# Identify number of records and features for identity_df
num_records_identity = identity_df.shape[0]
num_features_identity = identity_df.shape[1]

print(f"Number of records in identity.csv: {num_records_identity}")
print(f"Number of features in identity.csv: {num_features_identity}")

# Check for target variable (risk_flag) in identity_df
if 'risk_flag' in identity_df.columns:
    print("Target variable 'risk_flag' found in identity.csv.")
else:
    print("Target variable 'risk_flag' not found in identity.csv.")

print("\nFirst 5 rows of identity.csv:")
display(identity_df.head())

print("\nInformation about identity.csv:")
identity_df.info()

Number of records in identity.csv: 144233
Number of features in identity.csv: 41
Target variable 'risk_flag' not found in identity.csv.

First 5 rows of identity.csv:


,TransactionID,id_01,id_02,id_03,id_04,id_05,id_06,id_07,id_08,id_09,...,id_31,id_32,id_33,id_34,id_35,id_36,id_37,id_38,DeviceType,DeviceInfo
0,2987004,0.0,70787.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,samsung browser 6.2,32.0,2220x1080,match_status:2,T,F,T,T,mobile,SAMSUNG SM-G892A Build/NRD90M
1,2987008,-5.0,98945.0,NaN,NaN,0.0,-5.0,NaN,NaN,NaN,...,mobile safari 11.0,32.0,1334x750,match_status:1,T,F,F,T,mobile,iOS Device
2,2987010,-5.0,191631.0,0.0,0.0,0.0,0.0,NaN,NaN,0.0,...,chrome 62.0,NaN,NaN,NaN,F,F,T,T,desktop,Windows
3,2987011,-5.0,221832.0,NaN,NaN,0.0,-6.0,NaN,NaN,NaN,...,chrome 62.0,NaN,NaN,NaN,F,F,T,T,desktop,NaN
4,2987016,0.0,7460.0,0.0,0.0,1.0,0.0,NaN,NaN,0.0,...,chrome 62.0,24.0,1280x800,match_status:2,T,F,T,T,desktop,MacOS



Information about identity.csv:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 144233 entries, 0 to 144232
Data columns (total 41 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   TransactionID  144233 non-null  int64  
 1   id_01          144233 non-null  float64
 2   id_02          140872 non-null  float64
 3   id_03          66324 non-null   float64
 4   id_04          66324 non-null   float64
 5   id_05          136865 non-null  float64
 6   id_06          136865 non-null  float64
 7   id_07          5155 non-null    float64
 8   id_08          5155 non-null    float64
 9   id_09          74926 non-null   float64
 10  id_10          74926 non-null   float64
 11  id_11          140978 non-null  float64
 12  id_12          144233 non-null  object 
 13  id_13          127320 non-null  float64
 14  id_14          80044 non-null   float64
 15  id_15          140985 non-null  object 
 16  id_16          129340 non-null  object 
 

### Business Meaning of Columns
To fully understand the business meaning of columns, a data dictionary is required. Since it's not provided, I will proceed with the assumption that `risk_flag` is the target variable based on the instructions. The `df.info()` output above provides data types and non-null counts, which helps in initial understanding of the column characteristics.

From the head and info of both dataframes, we can infer some column meanings:

**`transactions.csv`**:
*   `TransactionID`: Unique identifier for each transaction.
*   `isFraud`: Likely indicates if a transaction is fraudulent (this would be a target variable if not `risk_flag`).
*   `TransactionDT`: Timestamp of the transaction.
*   `TransactionAmt`: Amount of the transaction.
*   `ProductCD`: Product code.
*   `card1` to `card6`: Anonymized card features.
*   `addr1`, `addr2`: Anonymized address features.
*   `P_emaildomain`, `R_emaildomain`: Primary and secondary email domains.
*   `C1` to `C14`: Anonymized counting features.
*   `D1` to `D15`: Anonymized timedelta features.
*   `M1` to `M9`: Anonymized matching features.
*   `V1` to `V339`: Anonymized Vesta engineered features.

**`identity.csv`**:
*   `TransactionID`: Unique identifier for each transaction (common key).
*   `DeviceType`, `DeviceInfo`: Information about the device used for the transaction.
*   `id_01` to `id_38`: Anonymized identity features.

**Target Variable:** As per the instructions, `risk_flag` is the target variable. I will assume this is present in one of the datasets or will be derived. Based on the `transactions.csv` content, `isFraud` seems like the most likely candidate for `risk_flag` if it's not explicitly named `risk_flag`.

## Task 1.2 Data Joining Strategy

### Identify Common Keys and Join Type

**Common Key:**
Both `transactions_df` and `identity_df` contain a column named `TransactionID`. This column appears to be a unique identifier for each transaction and is present in both datasets, making it the ideal common key for merging.

**Join Type and Justification:**

I will use a **left join** (or `merge` with `how='left'`) to combine `identity_df` with `transactions_df`.

*   **`transactions_df` (left dataframe):** Contains 296,234 records, representing the main set of transactions.
*   **`identity_df` (right dataframe):** Contains 144,233 records, providing additional identity-related information for a subset of the transactions.

**Justification for Left Join:**

1.  **Retain all transaction records:** The primary goal is often to analyze all transactions, whether they have associated identity information or not. A left join ensures that all records from `transactions_df` are kept.
2.  **Enrichment:** Identity information from `identity_df` will be added to the matching transaction records. For transactions present in `transactions_df` but not in `identity_df`, the identity-related columns will have `NaN` (Not a Number) values, indicating missing identity information.
3.  **Preservation of context:** Since `transactions_df` is larger and likely represents the core dataset for fraud detection, keeping all its records is crucial for a comprehensive analysis.

This strategy allows us to analyze the impact of identity features on all transactions, while also being able to identify transactions where identity data is unavailable.

In [5]:
# Perform a left merge
merged_df = pd.merge(transactions_df, identity_df, on='TransactionID', how='left')

print("Merged DataFrame created successfully.")
print(f"Number of records in merged_df: {merged_df.shape[0]}")
print(f"Number of features in merged_df: {merged_df.shape[1]}")

print("\nFirst 5 rows of merged_df:")
display(merged_df.head())

print("\nInformation about merged_df (showing first few columns):")
merged_df.info(verbose=True, show_counts=True, max_cols=50)

Merged DataFrame created successfully.
Number of records in merged_df: 296234
Number of features in merged_df: 434

First 5 rows of merged_df:


,TransactionID,isFraud,TransactionDT,TransactionAmt,ProductCD,card1,card2,card3,card4,card5,...,id_31,id_32,id_33,id_34,id_35,id_36,id_37,id_38,DeviceType,DeviceInfo
0,2987000,0,86400,68.5,W,13926,NaN,150.0,discover,142.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2987001,0,86401,29.0,W,2755,404.0,150.0,mastercard,102.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2987002,0,86469,59.0,W,4663,490.0,150.0,visa,166.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2987003,0,86499,50.0,W,18132,567.0,150.0,mastercard,117.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2987004,0,86506,50.0,H,4497,514.0,150.0,mastercard,102.0,...,samsung browser 6.2,32.0,2220x1080,match_status:2,T,F,T,T,mobile,SAMSUNG SM-G892A Build/NRD90M



Information about merged_df (showing first few columns):
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 296234 entries, 0 to 296233
Data columns (total 434 columns):
 #    Column          Non-Null Count   Dtype  
---   ------          --------------   -----  
 0    TransactionID   296234 non-null  int64  
 1    isFraud         296234 non-null  int64  
 2    TransactionDT   296234 non-null  int64  
 3    TransactionAmt  296234 non-null  float64
 4    ProductCD       296234 non-null  object 
 5    card1           296234 non-null  int64  
 6    card2           291520 non-null  float64
 7    card3           296229 non-null  float64
 8    card4           296222 non-null  object 
 9    card5           294641 non-null  float64
 10   card6           296225 non-null  object 
 11   addr1           263339 non-null  float64
 12   addr2           263339 non-null  float64
 13   dist1           111361 non-null  float64
 14   dist2           23648 non-null   float64
 15   P_emaildomain   249473 no

## Task 2.1 Missing Value Analysis

In [6]:
# Calculate missing values and their percentages
missing_values = merged_df.isnull().sum()
missing_percentage = (merged_df.isnull().sum() / len(merged_df)) * 100

missing_info = pd.DataFrame({
    'Missing Count': missing_values,
    'Missing Percentage': missing_percentage
})

# Filter to show only columns with missing values and sort them
missing_info = missing_info[missing_info['Missing Count'] > 0].sort_values(by='Missing Percentage', ascending=False)

print("Columns with missing values and their percentages:")
display(missing_info)

Columns with missing values and their percentages:


,Missing Count,Missing Percentage
id_24,293264,98.997414
id_25,293028,98.917747
id_08,293021,98.915384
id_07,293021,98.915384
id_21,293018,98.914372
...,...,...
C11,1,0.000338
C10,1,0.000338
C12,1,0.000338
C13,1,0.000338


### Categorizing Missingness

Based on the analysis of missing values, we can broadly categorize the missingness:

*   **High Missingness (over 70-80%):** Many `V` columns (e.g., `V306` to `V339`, `V290` to `V305`, `V153` to `V166`), and some `id_XX` columns (e.g., `id_07`, `id_08`, `id_21` to `id_27`, `id_24`). These columns are likely to be dropped due to their sparsity.
    *   **Potential Reason:** This could be **Missing Not At Random (MNAR)** if the absence of data is related to the unobserved value itself (e.g., certain types of fraudulent activities might not leave digital traces captured by these features). It could also be that these features are only relevant or collected for a very specific subset of transactions/users.

*   **Moderate Missingness (around 30-70%):** Many other `V` columns, `dist1`, `dist2`, `D` columns (e.g., `D7`, `D8`), and some `id_XX` columns (e.g., `id_03`, `id_04`, `id_18`, `id_30`, `id_32`, `id_33`).
    *   **Potential Reason:** This could be **Missing At Random (MAR)**, where the missingness can be explained by other observed variables (e.g., `id_03` might be missing more often for specific `DeviceType`s). It could also be **MNAR** if the missingness is tied to the underlying fraud behavior.

*   **Low Missingness (under 30%):** `R_emaildomain`, `P_emaildomain`, `addr1`, `addr2`, `card2`, `card5`, and some `id_XX` columns (e.g., `id_02`, `id_05`, `id_06`, `id_09`, `id_10`, `id_11`, `id_13`, `id_14`, `id_15`, `id_16`, `id_17`, `id_19`, `id_20`, `id_28`, `id_29`, `id_31`, `id_34`, `id_35`, `id_36`, `id_37`, `id_38`, `DeviceType`, `DeviceInfo`). Also, `card3`, `card4`, `card6`, `C` columns, and `D1` have very few missing values.
    *   **Potential Reason:** This might be **Missing Completely At Random (MCAR)** due to random data entry errors or system glitches, or **MAR** if it's related to other observed features. These are typically easier to impute or handle without significant loss of information.

**Considerations for `id_XX` columns:** The identity information is only available for a subset of transactions. The missingness in these columns (for transactions not found in `identity_df`) is a structural missingness, related to the join operation. These are essentially `NaN` for transactions without identity records. Their missingness percentages (e.g., `id_01` at ~70%) reflect this and will be a major factor in how they are handled.

## Task 2.2 Data Cleaning

### Handling Missing Values: Dropping Columns with High Missingness

Based on the missing value analysis, several columns have an extremely high percentage of missing data (e.g., >90%). Keeping these columns is unlikely to add value to the model and can increase computational overhead. Therefore, I will drop columns where more than 90% of the values are missing.

In [7]:
# Identify columns to drop based on a threshold (e.g., 90% missing)
threshold = 90
columns_to_drop = missing_info[missing_info['Missing Percentage'] > threshold].index.tolist()

print(f"Number of columns to drop (>{threshold}% missing): {len(columns_to_drop)}")
print("Columns to be dropped:")
print(columns_to_drop)

# Drop the identified columns from merged_df
merged_df_cleaned = merged_df.drop(columns=columns_to_drop)

print(f"\nShape of DataFrame before dropping: {merged_df.shape}")
print(f"Shape of DataFrame after dropping high missingness columns: {merged_df_cleaned.shape}")

# Update missing_info for the cleaned DataFrame
missing_values_cleaned = merged_df_cleaned.isnull().sum()
missing_percentage_cleaned = (merged_df_cleaned.isnull().sum() / len(merged_df_cleaned)) * 100

missing_info_cleaned = pd.DataFrame({
    'Missing Count': missing_values_cleaned,
    'Missing Percentage': missing_percentage_cleaned
})

# Filter to show only columns with remaining missing values and sort them
missing_info_cleaned = missing_info_cleaned[missing_info_cleaned['Missing Count'] > 0].sort_values(by='Missing Percentage', ascending=False)

print("\nRemaining columns with missing values and their percentages after dropping high missingness columns:")
display(missing_info_cleaned.head(20)) # Display top 20 to check

Number of columns to drop (>90% missing): 12
Columns to be dropped:
['id_24', 'id_25', 'id_08', 'id_07', 'id_21', 'id_26', 'id_23', 'id_27', 'id_22', 'D7', 'dist2', 'id_18']

Shape of DataFrame before dropping: (296234, 434)
Shape of DataFrame after dropping high missingness columns: (296234, 422)

Remaining columns with missing values and their percentages after dropping high missingness columns:


,Missing Count,Missing Percentage
D13,265654,89.677080
D14,264088,89.148443
D12,262259,88.531026
D6,257879,87.052465
id_03,257181,86.816841
id_04,257181,86.816841
D8,251436,84.877495
id_10,251436,84.877495
D9,251436,84.877495
id_09,251436,84.877495


### Further Handling Missing Values: Dropping Columns with over 70% Missingness

Although columns with >90% missing values were dropped, many still remain with significant missingness (e.g., 70-89%). To create a more robust dataset for modeling and reduce the burden of imputation, I will apply a stricter threshold and drop all columns with more than **70% missing values**.

In [8]:
# Identify columns to drop based on a new threshold (e.g., 70% missing)
threshold_2 = 70
columns_to_drop_2 = missing_info_cleaned[missing_info_cleaned['Missing Percentage'] > threshold_2].index.tolist()

print(f"Number of additional columns to drop (>{threshold_2}% missing): {len(columns_to_drop_2)}")
print("Additional columns to be dropped:")
print(columns_to_drop_2)

# Drop the identified columns from the already cleaned DataFrame
merged_df_cleaned_2 = merged_df_cleaned.drop(columns=columns_to_drop_2)

print(f"\nShape of DataFrame before this step: {merged_df_cleaned.shape}")
print(f"Shape of DataFrame after dropping columns with >{threshold_2}% missingness: {merged_df_cleaned_2.shape}")

# Update missing_info for the newly cleaned DataFrame
missing_values_final = merged_df_cleaned_2.isnull().sum()
missing_percentage_final = (merged_df_cleaned_2.isnull().sum() / len(merged_df_cleaned_2)) * 100

missing_info_final = pd.DataFrame({
    'Missing Count': missing_values_final,
    'Missing Percentage': missing_percentage_final
})

# Filter to show only columns with remaining missing values and sort them
missing_info_final = missing_info_final[missing_info_final['Missing Count'] > 0].sort_values(by='Missing Percentage', ascending=False)

print("\nRemaining columns with missing values and their percentages after dropping columns with >70% missingness:")
display(missing_info_final.head(30)) # Display top 30 to check

# Store the final cleaned DataFrame for subsequent steps
final_df = merged_df_cleaned_2

Number of additional columns to drop (>70% missing): 199
Additional columns to be dropped:
['D13', 'D14', 'D12', 'D6', 'id_03', 'id_04', 'D8', 'id_10', 'D9', 'id_09', 'id_33', 'id_30', 'id_32', 'id_34', 'id_14', 'V138', 'V165', 'V166', 'V149', 'V150', 'V151', 'V152', 'V164', 'V163', 'V162', 'V161', 'V156', 'V155', 'V154', 'V153', 'V160', 'V159', 'V158', 'V157', 'V141', 'V142', 'V143', 'V144', 'V148', 'V147', 'V146', 'V145', 'V139', 'V140', 'V337', 'V333', 'V332', 'V338', 'V335', 'V336', 'V339', 'V329', 'V330', 'V327', 'V331', 'V326', 'V323', 'V322', 'V328', 'V334', 'V325', 'V324', 'M9', 'M8', 'M7', 'DeviceInfo', 'id_13', 'id_16', 'V262', 'V263', 'V264', 'V265', 'V266', 'V267', 'V268', 'V269', 'V273', 'V274', 'V275', 'V276', 'V277', 'V278', 'V260', 'V261', 'V218', 'V217', 'V224', 'V225', 'V223', 'V219', 'V231', 'V232', 'V229', 'V230', 'V233', 'V235', 'V228', 'V226', 'V247', 'V248', 'V240', 'V241', 'V242', 'V243', 'V244', 'V246', 'V253', 'V254', 'V249', 'V252', 'V257', 'V258', 'V237', 'V

,Missing Count,Missing Percentage
V8,186443,62.937745
V5,186443,62.937745
V6,186443,62.937745
V7,186443,62.937745
V1,186443,62.937745
V4,186443,62.937745
V3,186443,62.937745
V2,186443,62.937745
D11,186443,62.937745
V9,186443,62.937745


### Handling Remaining Missing Values and Justification

After dropping columns with high missingness, we are left with a more manageable set of missing values. For the remaining columns, we need to choose appropriate imputation strategies based on their data type and nature.

**Numerical Columns:**
*   **Strategy:** Impute with the **median** or **mean**. Median is generally preferred as it's less sensitive to outliers.
*   **Justification:** Preserves the existing data distribution and allows the use of these features in models. Creating new indicator variables for missingness can also be considered for numerical columns where 'missing' might convey specific information.

**Categorical (Object) Columns:**
*   **Strategy:** Impute with the **mode** (most frequent category) or a new category like **'missing'**.
*   **Justification:** Imputing with 'missing' as a new category is often a good strategy as it retains the information that the value was originally absent, which might be predictive. Using the mode can be suitable if the missingness is small and truly random.

#### Justification for dropping columns versus imputation:

*   **Dropping columns (e.g., >70% missing):** When a very large proportion of data is missing, the column often lacks sufficient information to be useful. Imputing such a large number of values, especially with simple strategies, can introduce significant bias and artificial patterns, potentially degrading model performance. The information content is too low to justify retention.

*   **Imputing remaining missing values:** For columns with a smaller percentage of missing data (e.g., <70%), imputation is a better strategy than dropping the entire column. The column still contains valuable information from the existing non-missing data. Imputation helps retain these features, allowing models to learn from them, without losing entire records. The choice of imputation method aims to minimize distortion of the original data's statistical properties.


### Preventing Target Leakage

Target leakage occurs when your training data includes information that would not be available at the time of prediction, leading to an overly optimistic model performance that doesn't generalize to new, unseen data.

In this dataset, the target variable is `isFraud` (which we are using in place of `risk_flag`). We need to ensure that none of the features directly encode information that would *only* be known after a transaction is confirmed as fraudulent.

**Considerations:**

*   **No Obvious Direct Leakage:** Based on the feature descriptions (`TransactionID`, `TransactionDT`, `TransactionAmt`, `cardX`, `addrX`, `P_emaildomain`, `R_emaildomain`, `CX`, `DX`, `MX`, `VX`, `id_XX`, `DeviceType`, `DeviceInfo`), there isn't an immediately obvious feature that directly 'tells' the model if `isFraud` is 1 *before* the fraud occurs.
*   **Time-based Features:** `TransactionDT` represents time. If we were to create features like 'number of frauds in the next X minutes/hours' using future data, that would be leakage. We must ensure any time-based features are derived only from past information relative to `TransactionDT` or properties of `TransactionDT` itself (e.g., day of week, hour of day).
*   **Feature Engineering:** When engineering new features, extreme care must be taken. For example, if a feature is created that aggregates information *after* a transaction, and that aggregation includes the target, it would lead to leakage.

**Current Approach:**

For now, we will proceed with the existing features. Any feature engineering performed later will explicitly consider and avoid target leakage by using only information available at the time of the transaction. The `isFraud` column itself will be separated as the target variable for model training and will not be used as a feature.

### Imputing Remaining Missing Values

Now, we will impute the remaining missing values in `final_df` using the strategies outlined: median for numerical columns and mode for categorical columns.

In [9]:
# Separate numerical and categorical columns
numerical_cols = final_df.select_dtypes(include=['int64', 'float64']).columns
categorical_cols = final_df.select_dtypes(include=['object']).columns

print("Imputing numerical columns with median...")
for col in numerical_cols:
    if final_df[col].isnull().any():
        median_val = final_df[col].median()
        final_df[col].fillna(median_val, inplace=True)

print("Imputing categorical columns with mode...")
for col in categorical_cols:
    if final_df[col].isnull().any():
        # Mode might return multiple values if there's a tie, so take the first one
        mode_val = final_df[col].mode()[0]
        final_df[col].fillna(mode_val, inplace=True)

print("Missing value imputation complete.")

# Verify no more missing values
missing_after_imputation = final_df.isnull().sum().sum()

if missing_after_imputation == 0:
    print("\nSuccessfully imputed all missing values. The DataFrame now has no missing data.")
else:
    print(f"\nWarning: {missing_after_imputation} missing values still remain after imputation.")

print(f"Final DataFrame shape: {final_df.shape}")
print("First 5 rows of the final cleaned DataFrame:")
display(final_df.head())

Imputing numerical columns with median...


/tmp/ipykernel_16417/827801366.py:9: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  final_df[col].fillna(median_val, inplace=True)


Imputing categorical columns with mode...


/tmp/ipykernel_16417/827801366.py:16: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  final_df[col].fillna(mode_val, inplace=True)


Missing value imputation complete.

Successfully imputed all missing values. The DataFrame now has no missing data.
Final DataFrame shape: (296234, 223)
First 5 rows of the final cleaned DataFrame:


,TransactionID,isFraud,TransactionDT,TransactionAmt,ProductCD,card1,card2,card3,card4,card5,...,V312,V313,V314,V315,V316,V317,V318,V319,V320,V321
0,2987000,0,86400,68.5,W,13926,369.0,150.0,discover,142.0,...,0.0,0.0,0.0,0.0,0.0,117.0,0.0,0.0,0.0,0.0
1,2987001,0,86401,29.0,W,2755,404.0,150.0,mastercard,102.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,2987002,0,86469,59.0,W,4663,490.0,150.0,visa,166.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,2987003,0,86499,50.0,W,18132,567.0,150.0,mastercard,117.0,...,135.0,0.0,0.0,0.0,50.0,1404.0,790.0,0.0,0.0,0.0
4,2987004,0,86506,50.0,H,4497,514.0,150.0,mastercard,102.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
